File: null_object/java/CarRental.java
interface Veicle {

	default void start() {
		System.out.println("Starting veicle");
	}

	default void stop() {
		System.out.println("Stoping veicle");
	}

	default void accelerate() {
		System.out.println("Accelerating veicle");
	}

	default void park() {
		System.out.println("Parking veicle");
	}
}


class Car implements Veicle {

	String color;
	String model;
	String brand;

	Car(String color, String model, String brand) {
		this.color = color;
		this.model = model;
		this.brand = brand;
	}

	public String getColor() {
		return color;
	}

	public String getModel() {
		return model;
	}

	public String getBrand() {
		return brand;
	}
}

final class NullCar extends Car {

	public NullCar() {
		super("Black", "A8", "Audi");
	}

	@Override
	public String getColor() {
		return super.getColor();
	}

	@Override
	public String getModel() {
		return super.getModel();
	}

	@Override
	public String getBrand() {
		return super.getBrand();
	}
}

final class CarRental {

	Car rental() {
		return new NullCar();
	}


	public static void main(String[] args) {
		final Car car = new CarRental().rental();
		System.out.printf("%s %s %s\n", car.getBrand(), car.getModel(), car.getColor());
		car.start();
		car.park();
	}
}

File: null_object/kotlin/car_rental.kt
interface Veicle {

	fun start(): Unit {
		System.out.println("Starting veicle");
	}

	fun stop(): Unit {
		System.out.println("Stoping veicle");
	}

	fun accelerate(): Unit {
		System.out.println("Accelerating veicle");
	}

	fun park(): Unit {
		System.out.println("Parking veicle");
	}
}

open class Car(
	val brand: String,
	val model: String,
	val color: String
) : Veicle

class NullCar : Car("Audi", "A8", "Grey")


class CarRental {

	fun rental(): Car = NullCar()
}

fun main(): Unit {
	val carRental: CarRental = CarRental()
	val myCar: Car = carRental.rental()

	println("${myCar.brand} ${myCar.model} ${myCar.color}")

	myCar.accelerate()
}